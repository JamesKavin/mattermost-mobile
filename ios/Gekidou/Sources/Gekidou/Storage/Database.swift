//
//  Database.swift
//  Gekidou
//
//  Created by Miguel Alatzar on 8/20/21.
//

import Foundation
import SQLite3
import SQLite

// TODO: This should be exposed to Objective-C in order to handle
// any Database throwable methods.
enum DatabaseError: Error {
    case OpenFailure(_ dbPath: String)
    case MultipleServers
    case NoResults(_ query: String)
    case NoDatabase(_ serverUrl: String)
    case InsertError(_ statement: String)
}

extension DatabaseError: LocalizedError {
    var errorDescription: String? {
        switch self {
        case .OpenFailure(dbPath: let dbPath):
            return "Error opening database: \(dbPath)"
        case .MultipleServers:
            return "Cannot determine server URL as there are multiple servers"
        case .NoResults(query: let query):
            return "No results for query: \(query)"
        case .NoDatabase(serverUrl: let serverUrl):
            return "No database for server: \(serverUrl)"
        case .InsertError(statement: let statement):
            return "Insert error: \(statement)"
        }
    }
}

public class Database: NSObject {
    internal let DEFAULT_DB_NAME = "app.db"
    internal var DEFAULT_DB_PATH: String
    internal var defaultDB: OpaquePointer? = nil
    
    internal var serversTable = Table("Servers")
    internal var systemTable = Table("System")
    internal var teamTable = Table("Team")
    internal var myTeamTable = Table("MyTeam")
    internal var channelTable = Table("Channel")
    internal var channelInfoTable = Table("ChannelInfo")
    internal var channelMembershipTable = Table("ChannelMembership")
    internal var myChannelTable = Table("MyChannel")
    internal var myChannelSettingsTable = Table("MyChannelSettings")
    internal var postTable = Table("Post")
    internal var postsInChannelTable = Table("PostsInChannel")
    internal var postsInThreadTable = Table("PostsInThread")
    internal var postMetadataTable = Table("PostMetadata")
    internal var reactionTable = Table("Reaction")
    internal var fileTable = Table("File")
    internal var emojiTable = Table("CustomEmoji")
    
    @objc public static let `default` = Database()
    
    override private init() {
        let appGroupId = Bundle.main.object(forInfoDictionaryKey: "AppGroupIdentifier") as! String
        let sharedDirectory = FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: appGroupId)!
        let databaseUrl = sharedDirectory.appendingPathComponent("databases/\(DEFAULT_DB_NAME)")
        
        DEFAULT_DB_PATH = databaseUrl.path
    }
    
    @objc public func getOnlyServerUrlObjc() -> String {
        do {
            return try getOnlyServerUrl()
        } catch {
            print(error)
            return ""
        }
    }
    
    public func getOnlyServerUrl() throws -> String {
        let db = try Connection(DEFAULT_DB_PATH)
        let url = Expression<String>("url")
        let query = serversTable.select(url)
        
        var serverUrl: String?
        for result in try db.prepare(query) {
            if (serverUrl != nil) {
                throw DatabaseError.MultipleServers
            }
            
            serverUrl = try result.get(url)
        }
        
        if (serverUrl != nil) {
            return serverUrl!
        }
    
        throw DatabaseError.NoResults(query.asSQL())
    }
    
    internal func getDatabaseForServer(_ serverUrl: String) throws -> Connection {
        let db = try Connection(DEFAULT_DB_PATH)
        let url = Expression<String>("url")
        let dbPath = Expression<String>("db_path")
        let query = serversTable.select(dbPath).where(url == serverUrl)
        
        if let result = try db.pluck(query) {
            let path = try result.get(dbPath)
            return try Connection(path)
        }
        
        throw DatabaseError.NoResults(query.asSQL())
    }
    
    internal func queryCurrentUserId(_ serverUrl: String) throws -> String {
        let db = try getDatabaseForServer(serverUrl)
        
        let idCol = Expression<String>("id")
        let valueCol = Expression<String>("value")
        let query = systemTable.where(idCol == "currentUserId")
        
        if let result = try db.pluck(query) {
            return try result.get(valueCol).replacingOccurrences(of: "\"", with: "")
        }
        
        throw DatabaseError.NoResults(query.asSQL())
    }
    
    private func json(from object:Any?) -> String? {
        guard let object = object, let data = try? JSONSerialization.data(withJSONObject: object, options: []) else {
            return nil
        }
        return String(data: data, encoding: String.Encoding.utf8)
    }
}
