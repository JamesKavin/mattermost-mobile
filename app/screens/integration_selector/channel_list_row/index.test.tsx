// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import Database from '@nozbe/watermelondb/Database';
import React from 'react';

import {Preferences} from '@constants';
import {renderWithEverything} from '@test/intl-test-helper';
import TestHelper from '@test/test_helper';

import ChannelListRow from '.';

describe('components/integration_selector/channel_list_row', () => {
    let database: Database;
    const channel: Channel = {
        id: '1',
        create_at: 1111,
        update_at: 1111,
        delete_at: 0,
        team_id: 'my team',
        type: 'O',
        display_name: 'channel',
        name: 'channel',
        header: 'channel',
        purpose: '',
        last_post_at: 1,
        total_msg_count: 1,
        extra_update_at: 1,
        creator_id: '1',
        scheme_id: null,
        group_constrained: null,
        shared: true,
    };
    beforeAll(async () => {
        const server = await TestHelper.setupServerDatabase();
        database = server.database;
    });

    it('should match snapshot with open channel icon', () => {
        const wrapper = renderWithEverything(
            <ChannelListRow
                id='1234'
                theme={Preferences.THEMES.denim}
                channel={channel}
                selected={false}
                selectable={false}
                enabled={true}
                testID='ChannelListRow'
                onPress={() => {
                    // noop
                }}
            >
                <br/>
            </ChannelListRow>,
            {database},
        );

        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it('should match snapshot with private channel icon', () => {
        channel.type = 'P';

        const wrapper = renderWithEverything(
            <ChannelListRow
                id='1234'
                theme={Preferences.THEMES.denim}
                channel={channel}
                selected={false}
                selectable={false}
                enabled={true}
                testID='ChannelListRow'
                onPress={() => {
                    // noop
                }}
            >
                <br/>
            </ChannelListRow>,
            {database},
        );

        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it('should match snapshot with delete_at filled in', () => {
        channel.delete_at = 1111;
        channel.shared = false;

        const wrapper = renderWithEverything(
            <ChannelListRow
                id='1234'
                theme={Preferences.THEMES.denim}
                channel={channel}
                testID='ChannelListRow'
                enabled={true}
                selectable={false}
                selected={false}
                onPress={() => {
                    // noop
                }}
            >
                <br/>
            </ChannelListRow>,
            {database},
        );

        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it('should match snapshot with shared filled in', () => {
        channel.delete_at = 0;
        channel.shared = true;

        const wrapper = renderWithEverything(
            <ChannelListRow
                id='1234'
                theme={Preferences.THEMES.denim}
                channel={channel}
                testID='ChannelListRow'
                enabled={true}
                selectable={false}
                selected={false}
                onPress={() => {
                    // noop
                }}
            >
                <br/>
            </ChannelListRow>,
            {database},
        );

        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it('should match snapshot with purpose filled in', () => {
        channel.purpose = 'My purpose';

        const wrapper = renderWithEverything(
            <ChannelListRow
                id='1234'
                theme={Preferences.THEMES.denim}
                channel={channel}
                testID='ChannelListRow'
                enabled={true}
                selectable={false}
                selected={false}
                onPress={() => {
                    // noop
                }}
            >
                <br/>
            </ChannelListRow>,
            {database},
        );

        expect(wrapper.toJSON()).toMatchSnapshot();
    });
});
