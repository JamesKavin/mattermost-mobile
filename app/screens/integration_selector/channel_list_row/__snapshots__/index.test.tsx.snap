// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/integration_selector/channel_list_row should match snapshot with delete_at filled in 1`] = `
<View
  style={
    {
      "flex": 1,
      "flexDirection": "row",
      "overflow": "hidden",
      "paddingHorizontal": 15,
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "height": 65,
        "opacity": 1,
      }
    }
    testID="ChannelListRow"
  >
    <View
      style={
        {
          "flex": 1,
          "flexDirection": "row",
        }
      }
      testID="1234"
    >
      <View
        style={
          {
            "flex": 1,
            "flexDirection": "column",
          }
        }
        testID="ChannelListRow.1234"
      >
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <Icon
            name="archive-outline"
            style={
              {
                "color": "#3f4350",
                "fontFamily": "OpenSans",
                "fontSize": 16,
                "fontWeight": "400",
                "lineHeight": 24,
              }
            }
          />
          <Text
            style={
              {
                "color": "#3f4350",
                "fontFamily": "OpenSans",
                "fontSize": 16,
                "fontWeight": "400",
                "lineHeight": 24,
                "marginLeft": 5,
              }
            }
            testID="ChannelListRow.display_name"
          >
            channel
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`components/integration_selector/channel_list_row should match snapshot with open channel icon 1`] = `
<View
  style={
    {
      "flex": 1,
      "flexDirection": "row",
      "overflow": "hidden",
      "paddingHorizontal": 15,
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "height": 65,
        "opacity": 1,
      }
    }
    testID="ChannelListRow"
  >
    <View
      style={
        {
          "flex": 1,
          "flexDirection": "row",
        }
      }
      testID="1234"
    >
      <View
        style={
          {
            "flex": 1,
            "flexDirection": "column",
          }
        }
        testID="ChannelListRow.1234"
      >
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <Icon
            name="circle-multiple-outline"
            style={
              {
                "color": "#3f4350",
                "fontFamily": "OpenSans",
                "fontSize": 16,
                "fontWeight": "400",
                "lineHeight": 24,
              }
            }
          />
          <Text
            style={
              {
                "color": "#3f4350",
                "fontFamily": "OpenSans",
                "fontSize": 16,
                "fontWeight": "400",
                "lineHeight": 24,
                "marginLeft": 5,
              }
            }
            testID="ChannelListRow.display_name"
          >
            channel
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`components/integration_selector/channel_list_row should match snapshot with private channel icon 1`] = `
<View
  style={
    {
      "flex": 1,
      "flexDirection": "row",
      "overflow": "hidden",
      "paddingHorizontal": 15,
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "height": 65,
        "opacity": 1,
      }
    }
    testID="ChannelListRow"
  >
    <View
      style={
        {
          "flex": 1,
          "flexDirection": "row",
        }
      }
      testID="1234"
    >
      <View
        style={
          {
            "flex": 1,
            "flexDirection": "column",
          }
        }
        testID="ChannelListRow.1234"
      >
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <Icon
            name="circle-multiple-outline"
            style={
              {
                "color": "#3f4350",
                "fontFamily": "OpenSans",
                "fontSize": 16,
                "fontWeight": "400",
                "lineHeight": 24,
              }
            }
          />
          <Text
            style={
              {
                "color": "#3f4350",
                "fontFamily": "OpenSans",
                "fontSize": 16,
                "fontWeight": "400",
                "lineHeight": 24,
                "marginLeft": 5,
              }
            }
            testID="ChannelListRow.display_name"
          >
            channel
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`components/integration_selector/channel_list_row should match snapshot with purpose filled in 1`] = `
<View
  style={
    {
      "flex": 1,
      "flexDirection": "row",
      "overflow": "hidden",
      "paddingHorizontal": 15,
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "height": 65,
        "opacity": 1,
      }
    }
    testID="ChannelListRow"
  >
    <View
      style={
        {
          "flex": 1,
          "flexDirection": "row",
        }
      }
      testID="1234"
    >
      <View
        style={
          {
            "flex": 1,
            "flexDirection": "column",
          }
        }
        testID="ChannelListRow.1234"
      >
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <Icon
            name="circle-multiple-outline"
            style={
              {
                "color": "#3f4350",
                "fontFamily": "OpenSans",
                "fontSize": 16,
                "fontWeight": "400",
                "lineHeight": 24,
              }
            }
          />
          <Text
            style={
              {
                "color": "#3f4350",
                "fontFamily": "OpenSans",
                "fontSize": 16,
                "fontWeight": "400",
                "lineHeight": 24,
                "marginLeft": 5,
              }
            }
            testID="ChannelListRow.display_name"
          >
            channel
          </Text>
        </View>
        <Text
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            {
              "color": "rgba(63,67,80,0.5)",
              "fontFamily": "OpenSans",
              "fontSize": 14,
              "fontWeight": "400",
              "lineHeight": 20,
              "marginTop": 7,
            }
          }
        >
          My purpose
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`components/integration_selector/channel_list_row should match snapshot with shared filled in 1`] = `
<View
  style={
    {
      "flex": 1,
      "flexDirection": "row",
      "overflow": "hidden",
      "paddingHorizontal": 15,
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "height": 65,
        "opacity": 1,
      }
    }
    testID="ChannelListRow"
  >
    <View
      style={
        {
          "flex": 1,
          "flexDirection": "row",
        }
      }
      testID="1234"
    >
      <View
        style={
          {
            "flex": 1,
            "flexDirection": "column",
          }
        }
        testID="ChannelListRow.1234"
      >
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <Icon
            name="circle-multiple-outline"
            style={
              {
                "color": "#3f4350",
                "fontFamily": "OpenSans",
                "fontSize": 16,
                "fontWeight": "400",
                "lineHeight": 24,
              }
            }
          />
          <Text
            style={
              {
                "color": "#3f4350",
                "fontFamily": "OpenSans",
                "fontSize": 16,
                "fontWeight": "400",
                "lineHeight": 24,
                "marginLeft": 5,
              }
            }
            testID="ChannelListRow.display_name"
          >
            channel
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
